PRCS/prc-2.md

```diff
@@ -0,0 +1,174 @@
+    /// @dev Reverts if `tokens` length is not equal to `tokenIds` length.
+    /// Reverts if msg.sender is neither `owner` nor `operator` of LockInfo struct for
+    /// any of the input tokens.
+    /// Reverts if `unlockTime` of LockInfo struct for any of the input tokens is
+    /// either 0 or greater than block.timestamp.
+    /// Modifies a `LockInfo` struct `{unlockTime: block.timestamp + lockTime}` for each `token => tokenId`
+    /// Emits `Unlock` event.
+    /// @param tokens NFT tokens contract addresses
+    /// @param tokenIds NFT tokens identifiers
+    function withdraw(address[] memory tokens, uint256[] memory tokenIds) external;
+
+    /// @notice Returns `lockTime`, which is the value that gets added to block.timestamp and saved as unlockTime
+    /// in the requestUnlock function.
+    /// @return The `lockTime` variable
+    function getLockTime() external view returns (uint256);
+
+    /// @notice Changes the `lockTime` variable.
+    /// @dev This function should be protected with appropriate access control mechanisms.
+    /// The new value should be checked against a sane upper limit constant, which if exceeded,
+    /// should cause a revert.
+    /// Emits `LockTimeUpdate` event.
+    /// @param newLockTime New lockTime value
+    function setLockTime(uint256 newLockTime) external;
+}
+```
+
+A Hololocker implementation MUST implement the IERC721Receiver interface to be able to receive ERC721 assets via `IERC721.safeTransferFrom`.
+It MUST initialize a lock similarly as in the `lock` function, and it MUST emit the `Lock` event
+
+```solidity
+interface IERC721Receiver {
+    /// @dev Whenever an {IERC721} `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
+    ///  by `operator` from `from`, this function is called.
+    ///  It must return its Solidity selector to confirm the token transfer.
+    ///  If any other value is returned or the interface is not implemented by the recipient, the transfer will be
+    ///  reverted.
+    ///  The selector can be obtained in Solidity with `IERC721Receiver.onERC721Received.selector`.
+    ///  Note: the contract address is always the message sender.
+    /// @param _operator The address which called `safeTransferFrom` function
+    /// @param _from The address which previously owned the token
+    /// @param _tokenId The NFT identifier which is being transferred
+    /// @param _data Additional data with no specified format
+    /// @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+    ///  unless throwing
+    function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes _data) external returns(bytes4);
+}
+```
+
+## Rationale
+
+The Hololocker contract is in its core a NFT staking contract. While there have been attempts at EIPs on this matter, such as [EIP-4987](https://eips.ethereum.org/EIPS/eip-4987), none gained enough traction and approval to be finalized.

```

Comment:
We mentioned this EIP, but we didn't mention why we didn't use their standard other than it not getting traction. In other words, why do we not use the same interface they proposed so that we're the first dApp that implements their standard?

There might be a good reason, and we should mention it. Same for EIP 4400 and EIP 4799

Reply:
After reading them through, EIP-4400 seems irrelevant, as it's an extension for ERC721 tokens, and EIP-4799 seems irrelevant as it's tracking the designated owner by mapping the source NFT to a target NFT. So both of these proposals are doing different things than what we are doing.

As for EIP-4987, it being a stagnant proposal is indeed the major reason why we didn't use it. There are some minor differences, e.g. we log one extra variable (operator) in the events, so if we were to use the EIP-4987, we would have to emit the standard log and then our log with operator value (or we could scrap the operator from the log, it doesn't have real use-cases yet). The standard also necessitates keeping track of user's tokens balances, which we don't need to do. Other than that, we could have used it.